### PR TITLE
Voting Key generation using typescript and multi file support

### PR DESCRIPTION
**Pull request by fboucquez**
Update:

This PR now supports both the original catapult tools' voting key file generation and the native typescript. The original is the one used by default. You can enable the new native typescript with the useExperimentalNativeVotingKeyGeneration: true custom preset.

I'm also working on multi-voting file support. The voting file metadata (public key, start, and end epoch) is loaded directly from the .dat files. 

Voting file upgrade idea:

1. Bootstrap knows the network "current" finalizationEpoch (by either a manually upgraded custom preset value for offline setup or asking the network for online)
2. If the node's current voting file has an end epoch close to the current epoch (close to expiring), --upgrade will create 'private_key_tree2.dat' that continues the current file. I defined "Close to expiring"  when the epoch is in the upper half of the voting file. If the file's epoch length is 360, close to expiring will be 180+. Our nodes are 180 epochs, current epoch is 104 so they are close to expiring.
3. When a new voting file is created, bootstrap will advise running the link command again. The link command detects what has changed and announces the required transactions (linking the new voting key). The voting key files metadata (public key, start, and end epochs) will be in addresses.yml so it can also be linked using the symbol CLI if desired. 

For the end-user, the upgrade process is pretty much the same:

```
$ npm install -g symbol-bootstrap
$ symbol-bootstrap start -p mainnet -a dual -c custom-preset.yml --upgrde
>>> New Voting File Created!
$ symbol-bootstrap link
```

We can also add a new command "upgradeVotingFiles" that does the --upgrade of voting files explicitly 

TODOs:
- Test how the node behaves when transitioning from one file to another
- Double-check if we need to unlink already expired voting keys in the network. I think there is a limit about how many public keys can be linked. We may only need the current voting key and the future voting key
- Double-check want to do with old voting key files. Do we need to remove them? If we need to remove them, we need to "shift them", if voting file 1 is removed, voting file 2 should be renamed to voting key 1, 3 to voting key 2, etc

Slack chat: https://nemgroup-workspace.slack.com/archives/C015AEQ9ZL2/p1616410897191200

Related to https://github.com/nemtech/symbol-bootstrap/issues/149

Original:
This is my first implementation of voting key file generation done in Javascript/Typescript. 

I think the code is ok, the main issue is that the SDK is really slow generating keys and signing data. The SDK uses [tweetnacl](https://www.npmjs.com/package/tweetnacl), which is small but slow. These are some numbers on my dev machine when creating the voting key file:

(votingKeyEndEpoch - votingKeyStartEpoch + 1 -> TIME)
- 1 epoch -> 50 MS
- 5 epochs -> 114 MS
- 1000 epochs -> 10 seconds
- 10000 epochs -> 1 minute 41 seconds 
- 26280 epochs -> 4 Minutes 25 seconds

@gimer , could you validate the logic in https://github.com/nemtech/symbol-bootstrap/blob/native_voting_key_file/src/service/VotingUtils.ts#L24?

As unit testing, I'm comparing files generates with catapult tools against files generated using the native implementation. Checking that the headers and sizes are the same. I would add more assertions validating the internal items (private key and signatures). 

To speed up the implementation we could:

- Change the SDK crypto library. It could make the SDK faster but it would make it heavier, it may not work on Web, it could have secondary effects. Not Recommended at this stage.
- Change the crypto lib Bootstrap uses for the voting key implementation but keep the SDK for the rest of the key generation. Maybe something like [monocypher](https://monocypher.org/speed)

Before doing any improvement, I want to double-check the logic is fine. 

I'm pushing hard for this feature not because I want to reimplement the tool. This will allow me to make Bootstrap config offline without docker and OS independent (catapult tools wouldn't need to be compiled and installed in the offline machine). Supernodes will have fewer excuses to not use Bootstrap to setup the configuration (when runing barebone or with docker compose in the final online node machine)

Fixes https://github.com/nemtech/symbol-bootstrap/issues/169
Related to https://github.com/nemtech/symbol-bootstrap/issues/170

Update:
noble-ed25519 takes ~40 seconds for the 26280 epochs

### CODE REVIEW

File: src/service/VotingUtils.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+        //end-epoch (8b), / and each level holds this separately, so we left it as is
+        index = this.insert(result, Convert.numberToUint8Array(votingKeyEndEpoch, 8), index);
+        /// what follows are bound keys, there are (end - start + 1) of them.
+
+        // each key is:
+        for (let i = 0; i < items; i++) {
+            // random PRIVATE key (32b)
+            const randomPrivateKey = Crypto.randomBytes(32);
+            const randomKeyPar = KeyPair.createKeyPairFromPrivateKeyString(Convert.uint8ToHex(randomPrivateKey));
```
Comment (gimer) [inline src/service/VotingUtils.ts:58]: this looks weird or at least suspicious, doesn't KeyPair have ability to build the key straight from binary?

File: src/service/VotingUtils.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            // signature (64b)
+            // now the signature is usual signature done using ROOT private key on a following data:
+            //   (public key (32b), identifier (8b))
+            //
+            //   identifier is simply epoch, but, most importantly keys are written in REVERSE order.
+            //
+            //   i.e. say your start-epoch = 2, end-epoch = 42
+            const identifier = Convert.numberToUint8Array(votingKeyEndEpoch - i, 8);
+            const signature = KeyPair.sign(randomKeyPar, Uint8Array.from([...randomKeyPar.publicKey, ...identifier]));
+            index = this.insert(result, signature, index);
```
Comment (gimer) [inline src/service/VotingUtils.ts:69]: this is wrong, signature is done using ROOT priv key

**Comment by gimer** [comment]
`10000 `  - I hope you remember that for publicnet limit is 1460, so the tool should probably check that limit *somewhere*

File: src/service/VotingUtils.ts (hunk 1 above)
Comment (fboucquez) [reply src/service/VotingUtils.ts:58]: Yeap, notice that. It's in the SDK. It should be improved there

https://github.com/nemtech/symbol-sdk-typescript-javascript/blob/main/src/core/crypto/KeyPair.ts#L32

File: src/service/VotingUtils.ts (hunk 2 above)
Comment (fboucquez) [reply src/service/VotingUtils.ts]: Yeap, found it in my second commit. Thanks!

**Comment by fboucquez** [comment]
> 10000 - I hope you remember that for publicnet limit is 1460, so the tool should probably check that limit somewhere
Thanks. We also use bootstrap for private networks,  epoch length would be large in this case. I can add the limit the public/testnet nodes

**Comment by fboucquez** [comment]
Update, I've added multi implementation (all implementations pass the vector tests). The fastest one by far is noble-ed25519

File: src/service/VotingUtils.ts (hunk 1 above)
Comment (fboucquez) [reply src/service/VotingUtils.ts]: I've created https://github.com/nemtech/symbol-sdk-typescript-javascript/issues/770

File: src/service/VotingUtils.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            // signature (64b)
+            // now the signature is usual signature done using ROOT private key on a following data:
+            //   (public key (32b), identifier (8b))
+            //
+            //   identifier is simply epoch, but, most importantly keys are written in REVERSE order.
+            //
+            //   i.e. say your start-epoch = 2, end-epoch = 42
+            const identifier = Convert.numberToUint8Array(votingKeyEndEpoch - i, 8);
+            const signature = await this.implementation.sign(rootPrivateKey, Uint8Array.from([...randomKeyPar.publicKey, ...identifier]));
+            index = this.insert(result, signature, index);
```
Comment (gimer) [inline src/service/VotingUtils.ts:139]: looks ok to me

File: src/service/VotingUtils.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        result.set(value, index);
+        return index + value.length;
+    }
+
+    public async createVotingFile(
+        secret: string,
+        votingKeyStartEpoch: number,
+        votingKeyEndEpoch: number,
+        unitTestPrivateKeys: Uint8Array[] | undefined = undefined,
+    ): Promise<Uint8Array> {
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/service/VotingUtils.ts:91]: Some of the implementations were async, so the common denominator was Promise. I will keep the different implementation for now, but most likely reduce it to 1

File: src/service/VotingUtils.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    public static implementations = [VotingUtils.nobleImplementation, VotingUtils.tweetNaClImplementation, VotingUtils.forgeImplementation];
+
+    constructor(private readonly implementation: CryptoImplementation = VotingUtils.nobleImplementation) {}
+    public insert(result: Uint8Array, value: Uint8Array, index: number): number {
+        result.set(value, index);
+        return index + value.length;
+    }
+
+    public async createVotingFile(
+        secret: string,
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/service/VotingUtils.ts:87]: The secret is coming from the encrypted preset files or generated accounts. It's like the catapult tool that the secret param is a string

File: src/service/VotingUtils.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+        sign: async (keyPair: KeyPair, data: Uint8Array): Promise<Uint8Array> => {
+            const secretKey = new Uint8Array(64);
+            secretKey.set(keyPair.privateKey);
+            secretKey.set(keyPair.publicKey, 32);
+            return nacl.sign.detached(data, secretKey);
+        },
+    };
+
+    public static implementations = [VotingUtils.nobleImplementation, VotingUtils.tweetNaClImplementation, VotingUtils.forgeImplementation];
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/service/VotingUtils.ts:78]: To limit the impact, I've only used the noble implementation here. It seems pretty good and I'm keen to migrate the noble implementation, and maybe the whole voting key file generation to the SDK. This is for the future though. Probably replace tweetNaCl in the SDK
https://github.com/nemtech/symbol-sdk-typescript-javascript/issues/770

File: src/service/VotingService.ts [hunk 7]
```diff
@@ -14,108 +14,107 @@
...
-                `--endEpoch=${presetData.votingKeyEndEpoch}`,
-                `--output=/votingKeys/${privateKeyTreeFileName}`,
-            ];
-            await BootstrapUtils.deleteFolder(votingKeysFolder);
             await BootstrapUtils.mkdir(votingKeysFolder);
-            const binds = [`${votingKeysFolder}:/votingKeys:rw`];
+            const votingUtils = new VotingUtils();
+            await BootstrapUtils.deleteFile(join(votingKeysFolder, 'metadata.yml'));
+            let newFileCreated = false;
+            while (true) {
```
Comment (Wayonb) [inline src/service/VotingService.ts:54]: Not sure why ``while(true)``?  maybe I am missing something.
Are multiple files getting created?  I would assume when the loop is finish either it creates a voting key file or not.

File: src/service/VotingService.ts [hunk 8]
```diff
@@ -14,108 +14,107 @@
...
+                const epochs = votingKeyEndEpoch - votingKeyStartEpoch + 1;
+                logger.info(`Creating Voting key file of ${epochs} epochs for node ${nodeAccount.name}. This could take a while!`);
+                const privateKeyTreeFileName = `private_key_tree${currentVotingFiles.length + 1}.dat`;
+                if (presetData.useExperimentalNativeVotingKeyGeneration) {
+                    logger.info('Voting file is created using the native typescript voting key file generator!');
+                    const votingFile = await votingUtils.createVotingFile(votingPrivateKey, votingKeyStartEpoch, votingKeyEndEpoch);
+                    writeFileSync(join(votingKeysFolder, privateKeyTreeFileName), votingFile);
+                } else {
+                    logger.info(`Voting file is created using docker and the default's catapult.tools.votingkey`);
+                    const binds = [`${votingKeysFolder}:/votingKeys:rw`];
```
Comment (Wayonb) [inline src/service/VotingService.ts:82]: Should the votingService gets called here?

File: src/service/VotingUtils.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+import { Convert, Crypto } from 'symbol-sdk';
+import * as nacl from 'tweetnacl';
+
+const ed25519 = forge.pki.ed25519;
+export interface KeyPair {
+    privateKey: Uint8Array;
+    publicKey: Uint8Array;
+}
+export interface CryptoImplementation {
+    name: string;
```
Comment (Wayonb) [inline src/service/VotingUtils.ts:29]: Nice.  Good TS impl.

I think this file should be moved to a lib which can get used by Wallet or Symbol-cli.

**Review by Wayonb — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -277,6 +277,7 @@ General users should install this tool like any other node module.
 * [`symbol-bootstrap run`](docs/run.md) - It boots the network via docker using the generated `docker-compose.yml` file and configuration. The config and compose methods/commands need to be called before this method. This is just a wrapper for the `docker-compose up` bash call.
 * [`symbol-bootstrap start`](docs/start.md) - Single command that aggregates config, compose and run in one line!
 * [`symbol-bootstrap stop`](docs/stop.md) - It stops the docker-compose network if running (symbol-bootstrap started with --detached). This is just a wrapper for the `docker-compose down` bash call.
+* [`symbol-bootstrap upgradeVotingKeys`](docs/upgradeVotingKeys.md) - It upgrades the voting keys and files when required.
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply README.md:280]: Unsure, the command creates the file when necessary, it doesn't register in the network. For that, the link command is required (or symbol cli)

[the remark this answers is not in the document]
Comment (fboucquez) [reply README.md:280]: Unsure, the command creates the file when necessary, it doesn't register in the network. For that, the link command is required (or symbol cli)

File: presets/mainnet/network.yml [hunk 11]
```diff
@@ -27,6 +28,8 @@ importanceGrouping: 720
 votingSetGrouping: 1440
 minVotingKeyLifetime: 112
 maxVotingKeyLifetime: 360
+votingKeyDesiredLifetime: 360
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply presets/mainnet/network.yml:31]: Asking the community to upgrade every 3 months is too much I think. Especially when the "owner" of the node is not the admin, like in custodial setups. It's a balance between security and maintainability

File: src/service/VotingService.ts [hunk 12]
```diff
@@ -14,108 +14,107 @@
...
-                `--secret=${votingPrivateKey}`,
-                `--startEpoch=${presetData.votingKeyStartEpoch}`,
-                `--endEpoch=${presetData.votingKeyEndEpoch}`,
-                `--output=/votingKeys/${privateKeyTreeFileName}`,
-            ];
-            await BootstrapUtils.deleteFolder(votingKeysFolder);
             await BootstrapUtils.mkdir(votingKeysFolder);
-            const binds = [`${votingKeysFolder}:/votingKeys:rw`];
+            const votingUtils = new VotingUtils();
+            await BootstrapUtils.deleteFile(join(votingKeysFolder, 'metadata.yml'));
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/service/VotingService.ts:52]: Yes, this voting file metadata is not required anymore. Bootstrap now loads the information from the files directly, it knows how to read voting files.

File: src/service/VotingService.ts (hunk 7 above)
Comment (fboucquez) [reply src/service/VotingService.ts:54]: if you have a short votingKeyDesiredLifetime and a large votingKeyDesiredFutureLifetime in your custom preset, the service generates multiple files. Most people would not change this, but you have the option of having smaller files at once if you like

File: src/service/VotingService.ts (hunk 8 above)
Comment (fboucquez) [reply src/service/VotingService.ts:82]: I don't understand, could you rephrase?

File: src/service/VotingService.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/service/VotingService.ts:52]: I preferred to remove the unsued file instead of keeping it there.

File: src/service/LinkService.ts [hunk 13]
```diff
@@ -86,199 +95,302 @@ export class LinkService implements TransactionFactory {
...
+                voting: mainAccountInfo.supplementalPublicKeys.voting,
+            },
+            nodeAccount,
+            latestFinalizedBlockEpoch || presetData.lastKnownNetworkEpoch,
+            remoteTransactionFactory,
+            vrfTransactionFactory,
+            votingKeyTransactionFactory,
+        );
+        //Unlink transactions go first.
+        return transactions.sort((t1, t2) => t1.linkAction - t2.linkAction);
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/service/LinkService.ts:147]: It's what the sort is doing?

File: docs/upgradeVotingKeys.md [hunk 14]
```diff
@@ -0,0 +1,50 @@
+`symbol-bootstrap upgradeVotingKeys`
+====================================
+
+It upgrades the voting keys and files when required.
+
+Voting file upgrade:
+- If the node's current voting file has an end epoch close to the current epoch ("close to expiring") this command creates a new 'private_key_treeX.dat' that continues the current file.
+- "Close to expiring" happens when the epoch is in the upper half of the voting file. If the file's epoch length is 720, close to expiring will be 360+.
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply docs/upgradeVotingKeys.md:8]: I think at this stage allowing the user to upgrade 3 months in advance is fine, especially because this is the first upgrade and people need to get the process in place (like all nodes). We could reduce the default values in a later voting key upgrade

[the remark this answers is not in the document]
Comment (Wayonb) [reply docs/upgradeVotingKeys.md:8]: I think a month is more than enough time to run the update command.
Allnode could create a cron job that runs every 6 months to generate new key files.  If they want to run it early, then nothing stopping them since there is an override option.
We are kind thinking similar except your suggestion would require an update later to the bootstrap.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Wayonb) [reply README.md:280]: hehe...maybe ``register`` is not a good name either.
``upgradeVotingKeys`` doesnt make much sense to me since voting key/files are not upgradable. 
 @segfaultxavi thoughts?
 maybe ``addVotingKeys`` or ``createVotingKeys``

File: presets/mainnet/network.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (Wayonb) [reply presets/mainnet/network.yml:31]: yea, in the case where ``owner`` is not admin, I would think the owner is paying to secure their node.  They would want to follow best practices to secure the Symbol network which is in their best interest.
I am ok with leaving this.

File: src/service/LinkService.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (Wayonb) [reply src/service/LinkService.ts:147]: yea, I was saying its not obvious :)
I was saying it should be explicit.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (segfaultxavi) [reply README.md:280]: `renewVotingKeys`?

File: src/service/VotingService.ts (hunk 7 above)
Comment (Wayonb) [reply src/service/VotingService.ts:54]: ok.  I would suggest not generating keys that are not been used.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fboucquez) [reply README.md:280]: Options: 

1. updateVotingKeys
2. createVotingKeys
3. addVotingKeys
4. renewVotingKeys

maybe 2?

[the remark this answers is not in the document]
Comment (segfaultxavi) [reply README.md:280]: This command, together with `link`, does two things, right? Creates new keys AND removes (or unregisters) the old ones. I think `add` and `create` both fail to convey the second action. I can see @Wayonb's point that the keys are not actually "updated", so I suggested the `renew` option :)
But this command will be very seldom used (twice a year!) so let's go with whatever :D

File: docs/upgradeVotingKeys.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (fboucquez) [reply docs/upgradeVotingKeys.md:8]: Ok, I will upgrade the preset. Defaults will be:

votingKeyDesiredLifetime: 6 months in epochs (how long the file last)
votingKeyDesiredFutureLifetime: 1 month in epochs (1 month before the current file expires bootstrap will create a new file)

File: src/service/VotingService.ts (hunk 7 above)
Comment (fboucquez) [reply src/service/VotingService.ts:54]: Fair enough. I have removed the while and only one file can be created at a given time. Also, votingKeyDesiredFutureLifetime cannot be greater than votingKeyDesiredLifetime.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fboucquez) [reply README.md]: I've talked to @Wayonb , updateVotingKeys is the name.

File: src/commands/updateVotingKeys.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+import LoggerFactory from '../logger/LoggerFactory';
+import { BootstrapUtils, CommandUtils, ConfigLoader, CryptoUtils, RemoteNodeService, VotingService } from '../service';
+const logger: Logger = LoggerFactory.getLogger(LogType.System);
+
+export default class UpdateVotingKeys extends Command {
+    static description = `It updates the voting keys and files when required.
+
+Voting file update:
+- If the node's current voting file has an end epoch close to the current epoch ("close to expiring") this command creates a new 'private_key_treeX.dat' that continues the current file.
+- By default, "Close to expiring" happens the voting file in the . By default, bootstrap will allow creating new files once the current file reaches its last month.
```
Comment (segfaultxavi) [inline src/commands/updateVotingKeys.ts:29]: I don't know what are you trying to say here, the first sentence makes no sense.

**Review by segfaultxavi — CHANGES_REQUESTED** [review]

File: src/commands/updateVotingKeys.ts (hunk 15 above)
Comment (fboucquez) [reply src/commands/updateVotingKeys.ts]: Yeap, what about now?

**Review by segfaultxavi — DISMISSED** [review]
👍

**Review by Wayonb — DISMISSED** [review]
